FileStorageLevels/ReadLeveledStorage.cpp

```diff
@@ -0,0 +1,309 @@
+            }
+            // Less convenient case, the buffer ends somewhere beyond the next extent start, read to start and roll
+            auto res = update(buff, next->offset - position);
+            buff += res;
+            ret_val += res;
+            buff_size -= res;
+        }
+    }
+
+    return ret_val;
+}
+
+off_t ReadLeveledStorage::lseek(off_t offset, int whence)
+{
+    auto new_position = position;
+    switch(whence)
+    {
+    case SEEK_SET:
+        new_position = offset;
+        break;
+    case SEEK_CUR:
+        new_position += offset;
+        break;
+    case SEEK_END:
+        new_position = file_size + offset;
+        break;
+    default:
+        throw std::runtime_error("Unknown SEEK mode " + std::to_string(whence));
+    }
+
+    if(new_position < 0)
+    {
+        throw std::runtime_error("Invalid argument. Offset: " + std::to_string(offset) + " whence: " + std::to_string(whence));
+    }
+    position = new_position;
+    ::lseek(first_fd, position, SEEK_SET);
+    ::lseek(second_fd, position, SEEK_SET);
+    return position;
+}
+
+ssize_t ReadLeveledStorage::update(void* buff, size_t buff_size)
+{
+    auto read_res = ::read(second_fd, buff, buff_size);
+    ++stats.secondary_reads;
+    checkAndThrow(read_res);
+    if(read_res == 0)
+    {
+        return 0;
+    }
+
+    auto write_res = ::write(first_fd, buff, read_res);
+    checkAndThrow(write_res);
+    if(read_res != write_res)
+    {
+        throw std::logic_error("Read/write failed. The read operation yield " + std::to_string(read_res) +
+                               "bytes, but the write operation written " + std::to_string(write_res) + "bytes.");
+    }
+    if(position == 0 && extents.begin()->offset == 0 && extents.begin()->length == 0)
+    {
+        // delete convenience placeholder, since we have real data to populate it with.
+        extents.erase(extents.begin());
+    }
+
+    auto emplaced = extents.emplace(Extent{.offset = position, .length = static_cast<size_t>(write_res)});
+    if(!emplaced.second)
+    {
+        throw std::logic_error("Failed to emplace new extent - offset: " + std::to_string(emplaced.first->offset) +
+                               ", length: " + std::to_string(emplaced.first->length));
+    }
+    mergeAdjacent(emplaced.first);
+    position += write_res;
+    primed_file_size += write_res;
+    is_fully_primed = primed_file_size == file_size;
+    if(primed_file_size > file_size)
+    {
+        throw std::logic_error("Current primary file size " + std::to_string(primed_file_size) + " is greater than secondary file size " +
+                               std::to_string(file_size));
+    }
+    if(is_fully_primed)
+    {
+        extents.clear();
+        extents.emplace(Extent{.offset = 0, .length = file_size});
+    }
+    return write_res;
+}
+
+void ReadLeveledStorage::mergeAdjacent(Extents::const_iterator it)
+{
+    validateAdjacent(it);
+    {
+        auto prev = std::prev(it);
```

Comment:
Could it refer to extents.begin() here?

Reply:
Not sure if it could, but it is definitely worth checking